app/controllers/doorkeeper/tokens_controller.rb

```diff
@@ -88,7 +88,20 @@ def strategy
     end
 
     def authorize_response
-      @authorize_response ||= strategy.authorize
+      @authorize_response ||= begin
+        before_successful_authorization
+        auth = strategy.authorize
+        after_successful_authorization unless auth.is_a?(Doorkeeper::OAuth::ErrorResponse)
```

Comment:
Took a look at the code, yep, seems like `ErrorResponse` is a right solution (all the flows uses it for non-successful responses)